### PR TITLE
[MRG+2] First iteration towards using pytest

### PR DESCRIPTION
**Pull request by massich**
### Reference Issue
Works on #4412 

### Description
Make tests runnable with pytest without error and migrating as little as possible

### Todo:
- [x] Get pytest to work (whatever tests is running, JUST manage to get green stuff)
- [x] Compare the tests runned by pytest with those runned by nose tests. (Make them the same)
- [x] Update nosetests labels `'slow_test'` etc..
- [ ] ~Migrate Makefile~
- [ ] ~Review travis~

### CODE REVIEW

**Comment by dengemann** [comment]
@massich great! Let us know once you need some review. In the meantime you may want to consider some smaller issues for warming up the engine :) 🚗

File: .travis.yml [hunk 1]
```diff
@@ -10,72 +10,19 @@ env:
...
-        - os: osx
-          env: CONDA_DEPENDENCIES="numpy scipy matplotlib pandas scikit-learn h5py pillow statsmodels mayavi"
-               PIP_DEPENDENCIES="pysurfer nitime faulthandler joblib nibabel codecov nose-timer nose-faulthandler"
-               MNE_DIRS=". beamformer channels commands connectivity datasets decoding forward gui inverse_sparse io"
-        # OSX second half
-        - os: osx
-          env: CONDA_DEPENDENCIES="numpy scipy matplotlib pandas scikit-learn h5py pillow statsmodels mayavi"
-               PIP_DEPENDENCIES="pysurfer nitime faulthandler joblib nibabel codecov nose-timer nose-faulthandler"
-               MNE_DIRS="minimum_norm preprocessing realtime simulation stats time_frequency viz"
-
```
Comment (agramfort) [inline .travis.yml]: @massich you should not remove all these lines. They are here for a reason. Maybe just ask why. Just replace nosetests with pytest. It should be an isolated change

Comment (massich) [inline .travis.yml]: I've removed them just to reduce the time of travis while [WIP]. I guess I should had use a more informative commit message.

**Comment by agramfort** [comment]
Ok although the builds are split to reduce the testing time for each build config

File: mne/io/tests/test_pick.py [hunk 2]
```diff
@@ -8,7 +8,8 @@
 
 from mne import (pick_channels_regexp, pick_types, Epochs,
                  read_forward_solution, rename_channels,
-                 pick_info, pick_channels, __file__, create_info)
+                 pick_info, pick_channels, create_info)
+from mne import __file__ as my_file
```
Comment (massich) [inline mne/io/tests/test_pick.py]: I'm sure that `my_file` does not follow the naming convention. Suggestions are welcome. 

cc:@agramfort @dengemann

Comment (agramfort) [inline mne/io/tests/test_pick.py]: why do you need to make this change?

Comment (massich) [inline mne/io/tests/test_pick.py]: the mne __file__ refers to `mne-python/mne/__init__.py` and pytest collector confuses this path with `mne/io/tests/test_pick.py` file. Which if I'm not mistaken, is the value of __file__ when evaluating test_pick.py.

**Comment by massich** [comment]
I think this is a good start towards using pytest.

File: mne/io/tests/test_pick.py (hunk 2 above)
Comment (agramfort) [inline mne/io/tests/test_pick.py]: I would call it mne_root_init_fname but it's a test so no strong feeling.

File: mne/stats/tests/test_cluster_level.py [hunk 3]
```diff
@@ -406,7 +406,7 @@ def test_permutation_connectivity_equiv():
 
 
 @slow_test
-def spatio_temporal_cluster_test_connectivity():
+def test_spatio_temporal_cluster_connectivity():
```
Comment (agramfort) [inline mne/stats/tests/test_cluster_level.py]: nice catch.

File: .travis.yml [hunk 4]
```diff
@@ -1,4 +1,5 @@
 language: c
+dist: trusty
```
Comment (agramfort) [inline .travis.yml]: does it speed up things?

Comment (Eric89GXL) [inline .travis.yml]: I think it has been running on `trusty` lately already but this is at least explicit

Comment (massich) [inline .travis.yml]: @agramfort I don't think so. Maybe some minutes on each container. But it's not enough. I'll be looking at the tests in subsequent PRs.

**Comment by agramfort** [comment]
ok

**Comment by agramfort** [comment]
good to go from your end then?

File: Makefile [hunk 5]
```diff
@@ -39,44 +39,44 @@ testing_data:
...
+	$(PYTESTS) -a 'not slow_test' mne
 
 test-full: in
 	rm -f .coverage
-	$(NOSETESTS) mne
+	$(PYTESTS) mne
 
 test-no-network: in
-	sudo unshare -n -- sh -c 'MNE_SKIP_NETWORK_TESTS=1 nosetests mne'
+	sudo unshare -n -- sh -c 'MNE_SKIP_NETWORK_TESTS=1 py.test mne'
```
Comment (massich) [inline Makefile]: I don't understand what this is doing. I runned both nosetests and pytests changing this flag and there's no change in the test collection.

**Comment by agramfort** [comment]
it's a way to run the test mimicing a machine that cannot access the
network.

**Comment by Eric89GXL** [comment]
@massich can you rebase so that the tests are run again with #4458 merged?

**Comment by massich** [comment]
I think that only misses the circleci, which I don't understand why does not work. It's true that `plot_python_intro` misses the title and therefore it should fail. But what I don't understand why master is green.

**Comment by Eric89GXL** [comment]
>  I don't understand why does not work.

Upstream changes in Sphinx-gallery (more strict checks) that are fixed by #4468

**Comment by Eric89GXL** [comment]
#4468 is merged, if you rebase you should get all green (I don't think CircleCI builds a merged version)

**Comment by larsoner** [comment]
@massich ready to go from your end? If so I'll take a look and merge if it's all good.

**Comment by massich** [comment]
@larsoner go ahead. This pull was just to change the CI and give it a first go. Check that I didn't mess in the utils in the makefile. But I'm not quite sure how you guys use it. So if I didn't mimic the behavior 100%, let people will scream :) and we can improve it then.

File: .travis.yml [hunk 6]
```diff
@@ -12,69 +13,95 @@ matrix:
...
+               PIP_DEPENDENCIES="nitime joblib nibabel codecov nose-timer pytest-sugar pytest-cov pytest-attrib"
+               MNE_DIRS=". beamformer channels commands connectivity datasets decoding forward gui inverse_sparse io"
+               OPTION=""
+        # Py3k + non-default stim channel second half
+        - os: linux
+          env: PYTHON_VERSION=3.5 TEST_LOCATION=install MNE_STIM_CHANNEL=STI101
+               CONDA_DEPENDENCIES="numpy scipy matplotlib pandas scikit-learn h5py pillow statsmodels nose pytest"
+               PIP_DEPENDENCIES="nitime joblib nibabel codecov nose-timer pytest-sugar pytest-cov pytest-attrib"
+               MNE_DIRS="minimum_norm preprocessing realtime simulation stats time_frequency viz"
+               OPTION=""
```
Comment (larsoner) [inline .travis.yml]: Why are these duplicated (comment is the same as the one below)?

Comment (massich) [inline .travis.yml]: there were not actually equal. One was 3.5 the other 3.6. 
I've removed 3.6

**Comment by agramfort** [comment]
diff LGTM

can you just see what travis is not happy? thx @massich

**Comment by larsoner** [comment]
This one fails for the same reason JR's PR failed, sklearn API thing. I'll try to look into it

@agramfort feel free to merge if you're happy

**Comment by agramfort** [comment]
+1 for MRG when travis is happy

**Pull request merged**